src/writer/auth.py

```diff
@@ -35,12 +39,106 @@ class Auth:
                  unauthorized_action: Optional[Callable[[Request, Unauthorized], Response]] = None
     ):
         raise NotImplementedError
 
+@dataclasses.dataclass
+class BasicAuth(Auth):
+    """
+    Configure Writer Framework to use Basic Authentication. If this is set, Writer Framework will
+    ask anonymous users to authenticate using Basic Authentication.
+
+    >>> _auth = auth.BasicAuth(
+    >>>     login=os.getenv('LOGIN'),
+    >>>     password=os.getenv('PASSWORD')
+    >>> )
+    >>> writer.server.register_auth(_auth)
+
+    Brute force protection
+    ----------------------
+
+    A simple brute force protection is implemented by default. If a user fails to log in, the IP of this user is blocked.
+    Writer framework will ban the IP from either the `X-Forwarded-For` header or the `X-Real-IP` header or the client IP address.
+
+    When a user fails to log in, they wait 1 second before they can try again. This time can be modified by
+    modifying the value of `delay_after_failure`.
+
+    >>> _auth = auth.BasicAuth(
+    >>>     login=os.getenv('LOGIN'),
+    >>>     password=os.getenv('PASSWORD')
+    >>>     delay_after_failure=5 # 5 seconds delay after a failed login
+    >>> )
+    >>> writer.server.register_auth(_auth)
+
+    The user is stuck by default after a failure.
+
+    >>> _auth = auth.BasicAuth(
+    >>>     login=os.getenv('LOGIN'),
+    >>>     password=os.getenv('PASSWORD'),
+    >>>     delay_after_failure=5,
+    >>>     block_webserver_after_failure=False
+    >>> )
+    """
+    login: str
+    password: str
+    delay_after_failure: int = 1  # limit attempt when authentication fail (reduce brute force risk)
+    block_user_after_failure: bool = True  # delay the answer to the user after a failed login
+
+    callback_func: Optional[Callable[[Request, str, dict], None]] = None  # Callback to validate user authentication
+    unauthorized_action: Optional[Callable[[Request, Unauthorized], Response]] = None  # Callback to build its own page when a user is not allowed
+
+
+    def register(self,
+                 asgi_app: WriterFastAPI,
+                 callback: Optional[Callable[[Request, str, dict], None]] = None,
+                 unauthorized_action: Optional[Callable[[Request, Unauthorized], Response]] = None):
+
+        @asgi_app.middleware("http")
+        async def basicauth_middleware(request: Request, call_next):
+            import base64
+            client_ip = _client_ip(request)
+
+            try:
+                if client_ip in failed_attempts and time.time() - failed_attempts[client_ip] < self.delay_after_failure:
+                    remaining_time = int(self.delay_after_failure - (time.time() - failed_attempts[client_ip]))
+                    raise Unauthorized(status_code=429, message="Too Many Requests", more_info=f"You can try to log in every {self.delay_after_failure}s. Your next try is in {remaining_time}s.")
```

Comment:
Hey, are we reusing `Unauthorized` for both? Shouldn't something like `ClientError` be more descriptive? I got confused with the 429 using the Unauthorized template

Reply:
I decided to use the same exception because the authentication attempt is refused.

It could be an error from the client but also from the user. If he makes a mistake too many times, he reaches his maximum number of attempts in a given time. This can happen if the user persists so I prefer that they receive an explicit error message. A bit like when the user locks their phone because they entered the wrong PIN too many times.

I can create a ClientError exception which Unauthorized will inherit. Would that be more relevant?

Reply:
Makes sense, let's leave it as is.